cpp/src/arrow/compute/exec/asof_join_node.cc

```diff
@@ -459,17 +642,58 @@ class CompositeReferenceTable {
     if (!_ptr2ref.count((uintptr_t)ref.get())) _ptr2ref[(uintptr_t)ref.get()] = ref;
   }
 
-  template <class Builder, class PrimitiveType>
-  Result<std::shared_ptr<Array>> MaterializePrimitiveColumn(MemoryPool* memory_pool,
-                                                            size_t i_table,
-                                                            col_index_t i_col) {
-    Builder builder(memory_pool);
+  // this should really be a method on ArrayData
+  static bool IsNull(const std::shared_ptr<ArrayData>& source, row_index_t row) {
```

Comment:
I am surprised that we need this too. Maybe @westonpace knows existing method for this

Reply:
For some reason, `IsNull` does not exist in `ArrayData`, so I put it here for now; I'm fine moving it to `ArrayData` if reviewers accept it.

Reply:
An ``IsNull` method can be found in `ArraySpan`, but it doesn't make sense to create these objects in `AsofJoinNode` just for using `IsNull`.

Reply:
I'll move `IsNull` to `ArrayData`.